go-controller/pkg/cni/udn/primary_network_test.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+func TestWaitForPrimaryAnnotationFn(t *testing.T) {
+	tests := []struct {
+		description           string
+		nadName               string
+		annotationFromFn      *util.PodAnnotation
+		isReadyFromFn         bool
+		annotations           map[string]string
+		activeNetwork         *ovncnitypes.NetConf
+		getActiveNetworkError error
+		expectedIsReady       bool
+		expectedFound         bool
+		expectedAnnotation    *util.PodAnnotation
+		expectedNADName       string
+		expectedNetworkName   string
+		expectedMTU           int
+	}{
+		{
+			description:     "With non default nad should be ready",
+			nadName:         "red",
+			isReadyFromFn:   true,
+			expectedIsReady: true,
+		},
+		{
+			description: "With primary default should be ready",
+			nadName:     types.DefaultNetworkName,
+			annotationFromFn: &util.PodAnnotation{
+				Role: types.NetworkRolePrimary,
+			},
+			isReadyFromFn: true,
+			expectedAnnotation: &util.PodAnnotation{
+				Role: types.NetworkRolePrimary,
+			},
+			expectedIsReady: true,
+		},
+
+		{
+			description: "With missing primary annotation and active network should return not ready",
+			nadName:     types.DefaultNetworkName,
+			annotationFromFn: &util.PodAnnotation{
+				Role: types.NetworkRoleInfrastructure,
+			},
+			isReadyFromFn:   true,
+			expectedIsReady: false,
+		},
+		{
+			description: "With primary network annotation and missing active network should return not ready",
+			nadName:     types.DefaultNetworkName,
+			annotationFromFn: &util.PodAnnotation{
+				Role: types.NetworkRoleInfrastructure,
+			},
+			isReadyFromFn: true,
+			annotations: map[string]string{
+				"k8s.ovn.org/pod-networks": `{"ns1/nad1": {
+					"role": "primary",
+			        "mac_address": "0a:58:fd:98:00:01"
+				}}`,
+			},
+			expectedIsReady: false,
+		},
+		{
+			description: "With missing primary network annotation and active network should return not ready",
+			nadName:     types.DefaultNetworkName,
+			annotationFromFn: &util.PodAnnotation{
+				Role: types.NetworkRoleInfrastructure,
+			},
+			isReadyFromFn: true,
+			activeNetwork: &ovncnitypes.NetConf{
+				NetConf: cnitypes.NetConf{
+					Name: "blue",
+				},
+				NADName:  "ns1/nad1",
+				Topology: "layer2",
+			},
+			expectedIsReady: false,
+		},
+		{
+			description: "With primary network annotation and active network should return ready",
+			nadName:     types.DefaultNetworkName,
+			annotationFromFn: &util.PodAnnotation{
+				Role: types.NetworkRoleInfrastructure,
+			},
+			isReadyFromFn: true,
+			annotations: map[string]string{
+				"k8s.ovn.org/pod-networks": `{"ns1/nad1": {
+					"role": "primary",
+			        "mac_address": "0a:58:fd:98:00:01"
+				}}`,
+			},
+			activeNetwork: &ovncnitypes.NetConf{
+				NetConf: cnitypes.NetConf{
+					Name: "blue",
+				},
+				NADName:  "ns1/nad1",
+				Topology: "layer2",
+				MTU:      1300,
+			},
```

Comment:
didn't  we agree the active network would be a `NetInfo` ?

Reply:
This is just for testing, is way easier to compose the NetConf struct then the test convert it to proper NetInfo.

Reply:
I see. Maybe then rename this parameter ? From `activeNetwork` to something that hints this is actually input to build the active network from ? 

FWIW, I agree this is simpler.

Reply:
I don't know man this is active network and the active network is configured as a NAD, what we have here is the parsed version of the spec.conf which is better for testing purposes since it's structured.

Reply:
I agree with all that. but a CNI configuration object is not an active network. 

That's all I'm saying. But we can leave it as is, I won't block over this.

Reply:
just to illustrate with other unit test they are also using NetConf to create the NetInfo, since NetInfo is an interface you cannot directly construct it 
```golang
// setSecondaryNetworkTestData sets relevant test data (NAD, NetInfo & NB DB
    // initial data) assuming a secondary network of the given topoloy and      
    // subnet                                                                   
    setSecondaryNetworkTestData := func(topology, subnets string) {             
        nadNamespacedName = util.GetNADName(namespaceName1, nadName)            
        netconf := ovncnitypes.NetConf{                                         
            NetConf: cnitypes.NetConf{                                          
                Name: secondaryNetworkName,                                     
                Type: "ovn-k8s-cni-overlay",                                    
            },                                                                  
            Topology: topology,                                                 
            NADName:  nadNamespacedName,                                        
            Subnets:  subnets,                                                  
        }                                                                       
                                                                           
```